### DISCUSSION TITLE
Initializer Method Attribute

### DISCUSSION THREAD

**Discussion by TactiTac0z** [dotnet/csharplang#8527] [Language Ideas] [answered]
Sometimes it's cleaner to split your types initialization code into multiple methods, especially if your class is implemented in multiple files

I propose to add a "InstanceTypeInitializerMethodAttribute" (or perhaps "ConstructorMethodAttribute", etc name debatable) attribute that signifies that this instance method is to be called within the an instance constructor which allows this method to assign get only properties and readonly fields.
It would be an errors to use this method anywhere else. It could optionally (on a designlevel) be emitted with an unreferenceable name (like file accessibility types) so it is not unknowly used in reflection or unsupported languages (or versions)

Example:
```cs
class SomeType 
{
  public readonly int x;

  public SomeType()
  {
     InitializeX();
  }

  public void IllegalMethod()
  {
    // error CSxxxx: cannot use initializer method here
    InitializeX();
  }

  [InstanceTypeInitializerMethod]
  void InitializeX()
    => x = 1; // imagine a more complex initialization here
}
```

Right now, if someone wanted to use this structor, one would have to use unsafe code, avoiding readonly and get only or declaring `InitializeX` like `void InitializeX(out int x);` which can become very ugly very quickly.

If this has already been proposed before let me know.

**Comment by TahirAhmadov** [comment, marked as answer]
https://github.com/dotnet/csharplang/issues/4238
https://github.com/dotnet/csharplang/discussions/348
https://github.com/dotnet/csharplang/issues/1871
https://github.com/dotnet/csharplang/discussions/1296

**Comment by jaredpar** [comment]
I agree this is a valid scenario but the most likely way we'd end up addressing it is with `init` members. The `init` accessor already has the key properties here: can be called from ctor or object init and can modify `readonly` data. The original proposal considered that it could be applied to other members but LDM effectively said let's wait for the scenarios.
